### PR TITLE
[image] Factor out PNG processing

### PR DESCRIPTION
**Pull request by bertmaher**
*Description*: I'd like to use correctly-processed PNGs in a standalone utility, so I thought factoring this functionality out of the ImageClassifier executable would make sense.
*Testing*: ./tests/image/run.sh
*Documentation*: N/A
Part of #1976

### CODE REVIEW

File: include/glow/Base/Image.h [hunk 1]
```diff
@@ -40,6 +60,11 @@ bool writePngImage(Tensor *T, const char *filename,
                    std::pair<float, float> range,
                    bool useImagenetNormalization = false);
 
+Tensor readPngImageAndPreprocess(const std::string &filename,
```
Comment (jfix71) [inline include/glow/Base/Image.h]: nit: Doxygen!

File: tools/loader/ImageClassifier.cpp [hunk 2]
```diff
@@ -153,68 +120,46 @@ llvm::cl::opt<bool> useImagenetNormalization(
...
-  // Get first image's dimensions and check if grayscale or color.
-  size_t imgHeight, imgWidth;
-  bool isGray;
-  std::tie(imgHeight, imgWidth, isGray) = getPngInfo(filenames[0].c_str());
-  const size_t numChannels = isGray ? 1 : 3;
-
-  // N x C x H x W
-  inputImageData->reset(ElemKind::FloatTy,
-                        {numImages, numChannels, imgHeight, imgWidth});
+  Tensor initial =
```
Comment (jfix71) [inline tools/loader/ImageClassifier.cpp]: Curious, why did you decide to move loading the first image separately to set the dims of `inputImageData`? Before we just called `getPngInfo()` on the first image in order to set the dimensions of `inputImageData`, and then loaded all images in a loop. I guess it's one less `fopen()` this way? If you prefer this way then I think we could actually get rid of `getPngInfo()` entirely...

File: lib/Base/Image.cpp [hunk 3]
```diff
@@ -270,6 +286,48 @@ bool glow::writePngImage(Tensor *T, const char *filename,
...
+Tensor glow::readPngImageAndPreprocess(const std::string &filename,
+                                       ImageNormalizationMode imageNormMode,
+                                       ImageChannelOrder imageChannelOrder,
+                                       ImageLayout imageLayout,
+                                       bool useImagenetNormalization) {
+  // Get image dimensions and check if grayscale or color.
+  size_t imgHeight;
+  size_t imgWidth;
+  bool isGray;
+  std::tie(imgHeight, imgWidth, isGray) = getPngInfo(filename.c_str());
```
Comment (jfix71) [inline lib/Base/Image.cpp]: I don't think it's necessary to call `getPngInfo()` here -- you can get `imgHeight`, `imgWidth`, and `numChannels` from the shape of `imageData` below.

File: tools/loader/ImageClassifier.cpp [hunk 4]
```diff
@@ -153,68 +120,46 @@ llvm::cl::opt<bool> useImagenetNormalization(
...
-    inputImageData->transpose(&dataNHWC, NCHW2NHWC);
-    *inputImageData = std::move(dataNHWC);
-    break;
+  insertImage(IIDH, IH, 0);
+
+  for (size_t n = 1; n < filenames.size(); n++) {
+    Tensor localCopy = readPngImageAndPreprocess(filenames[n], imageNormMode,
+                                                 imageChannelOrder, imageLayout,
+                                                 useImagenetNormalization);
+    assert(std::equal(localCopy.dims().begin(), localCopy.dims().end(),
```
Comment (jfix71) [inline tools/loader/ImageClassifier.cpp]: nit: Can you do `localCopy.dims() == initial.dims()` instead?

File: include/glow/Base/Image.h (hunk 1 above)
Comment (nadavrot) [reply include/glow/Base/Image.h]: llvm::StringRef

File: tools/loader/ImageClassifier.cpp (hunk 4 above)
Comment (bertmaher) [reply tools/loader/ImageClassifier.cpp]: Hm, now that I'm using getPngInfo, I no longer have `initial`, so I'm using `std::equal` with `batch.dims().begin() + 1`.  What do you think of that?  It's kind of weird, but I don't have a better idea :-/

Comment (jfix71) [reply tools/loader/ImageClassifier.cpp]: You could create a const array above with the expected shape `{imgHeight, imgWidth, numChannels}` and then compare here. NBD either way though.

**Pull request merged**